File: leetcode-everyday/搜索算法/257-二叉树的所有路径/257-二叉树的所有路径.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode* node, string path, vector<string>& ans) {
        if(node != nullptr) {
            path += to_string(node->val);
            if(node->left == nullptr && node->right == nullptr) {
                ans.push_back(path);
            } else {
                path += "->";
                dfs(node->left, path, ans);
                dfs(node->right, path, ans);
            }
        }
    }
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        dfs(root, "", ans);
        return ans;
    }
};

File: leetcode-everyday/搜索算法/257-二叉树的所有路径/257-二叉树的所有路径.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
 var binaryTreePaths = function(root) {
  const ans = [];
  dfs(root, [], ans);
  return ans.map(item => item.join("->"));
};

function dfs(node, path, ans) {
  if(!node.left && !node.right) {
      ans.push([...path, node.val]);
      return
  }    
  if(node.left) {
      dfs(node.left, [...path, node.val], ans);
  }
  if(node.right) {
      dfs(node.right, [...path, node.val], ans);
  }    
}


/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
 var binaryTreePaths = function(root) {
  const ans = [];
  dfs(root, [], ans);
  return ans
};

function dfs(node, path, ans) {
  if(node) {
      path += node.val;
      if(!node.left && !node.right) {
          ans.push(path);
      } else {
          path += "->";
          dfs(node.left, path, ans);
          dfs(node.right, path, ans);
      }
  }
}